// @8base/web-auth0-auth-client/dist/WebAuth0AuthClient.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.WebAuth0AuthClient = void 0;

var _regenerator = _interopRequireDefault(require("@babel/runtime/regenerator"));

var _objectSpread2 = _interopRequireDefault(require("@babel/runtime/helpers/objectSpread"));

var _asyncToGenerator2 = _interopRequireDefault(require("@babel/runtime/helpers/asyncToGenerator"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _auth0Js = _interopRequireDefault(require("auth0-js"));

var R = _interopRequireWildcard(require("ramda"));

var localStorageAccessor = _interopRequireWildcard(require("./localStorageAccessor"));

var isEmptyOrNil = R.either(R.isNil, R.isEmpty);
var isEmailVerified = R.pipe(R.path(['idTokenPayload', 'email_verified']), R.equals(true));
var getEmail = R.path(['idTokenPayload', 'email']);
var getIdToken = R.path(['idToken']);
var getIdTokenPayload = R.prop('idTokenPayload');
var getState = R.pipe(R.prop('state'), function (state) {
  try {
    return JSON.parse(state);
  } catch (e) {
    return state;
  }
});
/**
 * Create instacne of the web auth0 auth client.
 * @param {string} workspaceId Identifier of the 8base app workspace.
 * @param {string} domain Domain. See auth0 documentation.
 * @param {string} clientId Client id. See auth0 documentation.
 * @param {string} redirectUri Redurect uri. See auth0 documentation.
 */

var WebAuth0AuthClient =
/*#__PURE__*/
function () {
  function WebAuth0AuthClient(_ref) {
    var _this = this;

    var domain = _ref.domain,
        clientId = _ref.clientId,
        redirectUri = _ref.redirectUri,
        workspaceId = _ref.workspaceId,
        logoutRedirectUri = _ref.logoutRedirectUri;
    (0, _classCallCheck2.default)(this, WebAuth0AuthClient);
    (0, _defineProperty2.default)(this, "auth0", void 0);
    (0, _defineProperty2.default)(this, "workspaceId", void 0);
    (0, _defineProperty2.default)(this, "logoutRedirectUri", void 0);
    (0, _defineProperty2.default)(this, "authorize",
    /*#__PURE__*/
    (0, _asyncToGenerator2.default)(
    /*#__PURE__*/
    _regenerator.default.mark(function _callee() {
      var options,
          _args = arguments;
      return _regenerator.default.wrap(function _callee$(_context) {
        while (1) {
          switch (_context.prev = _context.next) {
            case 0:
              options = _args.length > 0 && _args[0] !== undefined ? _args[0] : {};

              _this.auth0.authorize((0, _objectSpread2.default)({
                state: _this.getAuthorizeState()
              }, options));

            case 2:
            case "end":
              return _context.stop();
          }
        }
      }, _callee, this);
    })));
    (0, _defineProperty2.default)(this, "renewToken", function () {
      var options = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};
      return new Promise(function (resolve, reject) {
        _this.auth0.checkSession(options, function (error, result) {
          if (error) {
            reject(error || {});
            return;
          }

          resolve({
            state: getState(result),
            email: getEmail(result),
            idToken: getIdToken(result),
            isEmailVerified: isEmailVerified(result),
            idTokenPayload: getIdTokenPayload(result)
          });
        });
      });
    });
    (0, _defineProperty2.default)(this, "changePassword",
    /*#__PURE__*/
    (0, _asyncToGenerator2.default)(
    /*#__PURE__*/
    _regenerator.default.mark(function _callee2() {
      var _ref4, email;

      return _regenerator.default.wrap(function _callee2$(_context2) {
        while (1) {
          switch (_context2.prev = _context2.next) {
            case 0:
              _context2.next = 2;
              return _this.getAuthState();

            case 2:
              _ref4 = _context2.sent;
              email = _ref4.email;
              return _context2.abrupt("return", new Promise(function (resolve, reject) {
                _this.auth0.changePassword({
                  connection: 'Username-Password-Authentication',
                  email: email
                }, function (error) {
                  if (error) {
                    reject(error || {});
                    return;
                  }

                  resolve({
                    email: email
                  });
                });
              }));

            case 5:
            case "end":
              return _context2.stop();
          }
        }
      }, _callee2, this);
    })));
    (0, _defineProperty2.default)(this, "getAuthorizedData", function () {
      return new Promise(function (resolve, reject) {
        _this.auth0.parseHash(function (error, authResult) {
          if (error) {
            reject(error);
            return;
          }

          resolve({
            state: getState(authResult),
            email: getEmail(authResult),
            idToken: getIdToken(authResult),
            isEmailVerified: isEmailVerified(authResult),
            idTokenPayload: getIdTokenPayload(authResult)
          });
        });
      });
    });
    (0, _defineProperty2.default)(this, "setAuthState",
    /*#__PURE__*/
    function () {
      var _ref5 = (0, _asyncToGenerator2.default)(
      /*#__PURE__*/
      _regenerator.default.mark(function _callee3(state) {
        return _regenerator.default.wrap(function _callee3$(_context3) {
          while (1) {
            switch (_context3.prev = _context3.next) {
              case 0:
                localStorageAccessor.setAuthState(state);

              case 1:
              case "end":
                return _context3.stop();
            }
          }
        }, _callee3, this);
      }));

      return function (_x) {
        return _ref5.apply(this, arguments);
      };
    }());
    (0, _defineProperty2.default)(this, "getAuthState",
    /*#__PURE__*/
    (0, _asyncToGenerator2.default)(
    /*#__PURE__*/
    _regenerator.default.mark(function _callee4() {
      return _regenerator.default.wrap(function _callee4$(_context4) {
        while (1) {
          switch (_context4.prev = _context4.next) {
            case 0:
              return _context4.abrupt("return", localStorageAccessor.getAuthState());

            case 1:
            case "end":
              return _context4.stop();
          }
        }
      }, _callee4, this);
    })));
    (0, _defineProperty2.default)(this, "purgeAuthState",
    /*#__PURE__*/
    (0, _asyncToGenerator2.default)(
    /*#__PURE__*/
    _regenerator.default.mark(function _callee5() {
      return _regenerator.default.wrap(function _callee5$(_context5) {
        while (1) {
          switch (_context5.prev = _context5.next) {
            case 0:
              localStorageAccessor.purgeAuthState();

            case 1:
            case "end":
              return _context5.stop();
          }
        }
      }, _callee5, this);
    })));
    (0, _defineProperty2.default)(this, "checkIsAuthorized",
    /*#__PURE__*/
    (0, _asyncToGenerator2.default)(
    /*#__PURE__*/
    _regenerator.default.mark(function _callee6() {
      var _ref9, token;

      return _regenerator.default.wrap(function _callee6$(_context6) {
        while (1) {
          switch (_context6.prev = _context6.next) {
            case 0:
              _context6.next = 2;
              return _this.getAuthState();

            case 2:
              _ref9 = _context6.sent;
              token = _ref9.token;
              return _context6.abrupt("return", R.not(isEmptyOrNil(token)));

            case 5:
            case "end":
              return _context6.stop();
          }
        }
      }, _callee6, this);
    })));
    (0, _defineProperty2.default)(this, "logout",
    /*#__PURE__*/
    (0, _asyncToGenerator2.default)(
    /*#__PURE__*/
    _regenerator.default.mark(function _callee7() {
      var options,
          _args7 = arguments;
      return _regenerator.default.wrap(function _callee7$(_context7) {
        while (1) {
          switch (_context7.prev = _context7.next) {
            case 0:
              options = _args7.length > 0 && _args7[0] !== undefined ? _args7[0] : {};

              _this.auth0.logout((0, _objectSpread2.default)({
                returnTo: _this.logoutRedirectUri
              }, options));

            case 2:
            case "end":
              return _context7.stop();
          }
        }
      }, _callee7, this);
    })));
    this.workspaceId = workspaceId;
    this.logoutRedirectUri = logoutRedirectUri;
    this.auth0 = new _auth0Js.default.WebAuth({
      domain: domain,
      clientID: clientId,
      redirectUri: redirectUri,
      mustAcceptTerms: true,
      responseType: 'token id_token',
      scope: 'openid email profile',
      state: this.getAuthorizeState()
    });
  }

  (0, _createClass2.default)(WebAuth0AuthClient, [{
    key: "getAuthorizeState",
    value: function getAuthorizeState() {
      return this.workspaceId ? JSON.stringify({
        workspaceId: this.workspaceId
      }) : undefined;
    }
  }]);
  return WebAuth0AuthClient;
}();

exports.WebAuth0AuthClient = WebAuth0AuthClient;
